controllers/release/release_adapter.go

```diff
@@ -228,24 +229,36 @@ func (a *Adapter) EnsureSnapshotEnvironmentBindingIsCreated() (results.Operation
 // EnsureTargetContextIsSet is an operation that will ensure that the targetContext of the adapter is not nil. It will
 // set it to a context from the workspace in the ReleasePlan if one is defined or to its own context if not.
 func (a *Adapter) EnsureTargetContextIsSet() (results.OperationResult, error) {
-	if a.targetContext == nil {
+	// If the target namespace/workspace is not set we should get the target from the ReleasePlan
+	if a.release.Status.Target == (kcp.NamespaceReference{}) {
+		patch := client.MergeFrom(a.release.DeepCopy())
 		releasePlan, err := a.getReleasePlan()
 		if err != nil {
-			patch := client.MergeFrom(a.release.DeepCopy())
 			a.release.MarkInvalid(v1alpha1.ReleaseReasonReleasePlanValidationError, err.Error())
 			return results.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.release, patch))
 		}
-
-		if releasePlan.Spec.Target.Workspace == "" {
-			a.logger.Info("Running in a traditional kubernetes environment")
-			a.targetContext = a.context
-		} else {
+		if releasePlan.Spec.Target.Workspace != "" {
+			if _, ok := logicalcluster.ClusterFromContext(a.context); !ok {
```

Comment:
This check shouldn't trigger an error as we could be running in pre-kcp, which is fine.

Reply:
I have treat this error somewhere. The problem is if the`Workspace` is set and it is not a KCP cluster, it will trigger an untreated error in the controller. Although after the last change, I noticed it makes more sense to check this after I know that the Workspace at the `release.Status.Target` is already set. I just pushed it. Hope the next version is ok.

Reply:
yes, I agree this should be an error since it is after the check that workspace is defined

Reply:
Got you. I got confused about that.

Can we though reverse the if logic. Always better to start with the positive case (just a good practice)

Reply:
done.